### PR TITLE
Generate and add the validation field to CRD

### PR DESCRIPTION
**Pull request by ant31**
New binary to generate CRD used by the operator as yaml without deploying them.

```
$ po-crdgen --help
Usage of po-crdgen:
  -crd-apigroup string
    	prometheus CRD  API group name (default "monitoring.coreos.com")
  -crd-kinds value
    	customize CRD kind names
  -labels value
    	Labels to be add to all resources created by the operator
  -with-validation
    	Include the validation spec
```

```yaml
$ po-crdgen --crd-apigroup=foo.monitoring.coreos.com   
---
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  creationTimestamp: null
  name: prometheuses.foo.monitoring.coreos.com
spec:
  group: foo.monitoring.coreos.com
  names:
    kind: Prometheus
    plural: prometheuses
  scope: Namespaced
  version: v1
status:
  acceptedNames:
    kind: ""
    plural: ""
  conditions: null
---
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  creationTimestamp: null
  name: servicemonitors.foo.monitoring.coreos.com
spec:
  group: foo.monitoring.coreos.com
  names:
    kind: ServiceMonitor
    plural: servicemonitors
  scope: Namespaced
  version: v1
status:
  acceptedNames:
    kind: ""
    plural: ""
  conditions: null
---
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  creationTimestamp: null
  name: alertmanagers.foo.monitoring.coreos.com
spec:
  group: foo.monitoring.coreos.com
  names:
    kind: Alertmanager
    plural: alertmanagers
  scope: Namespaced
  version: v1
status:
  acceptedNames:
    kind: ""
    plural: ""
  conditions: null
```

### CODE REVIEW

File: cmd/po-crdgen/main.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+func init() {
+	cfg.CrdKinds = monitoringv1.DefaultCrdKinds
+	flagset := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	flagset.Var(&cfg.Labels, "labels", "Labels to be add to all resources created by the operator")
+	flagset.BoolVar(&cfg.EnableValidation, "with-validation", false, "Include the validation spec")
+	flagset.StringVar(&cfg.CrdGroup, "crd-apigroup", monitoringv1.Group, "prometheus CRD  API group name")
+	flagset.Var(&cfg.CrdKinds, "crd-kinds", "customize CRD kind names")
+	flagset.StringVar(&cfg.Namespace, "namespace", v1.NamespaceAll, "Namespace to scope the interaction of the Prometheus Operator and the apiserver.")
```
Comment (ant31) [inline cmd/po-crdgen/main.go]: Not used, to delete

File: cmd/po-crdgen/main.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2016 The prometheus-operator Authors
```
Comment (brancz) [inline cmd/po-crdgen/main.go]: new file new year

File: cmd/po-crdgen/crdgen.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package main
```
Comment (brancz) [inline cmd/po-crdgen/crdgen.go]: missing license header

**Comment by ant31** [comment]
@brancz should be good now. 
We have to double-check if the current 'json' annotation in the `types.go` are all correct or it will generate unwanted rules

**Comment by brancz** [comment]
@ant31 this needs a re-generate for tests to pass.

I'd say we can do a pre-release for validations as in `v0.17.0-beta` and ask the community to validate for us. There's no possibility for us to match the sheer number of examples the community will throw at it.

**Comment by brancz** [comment]
this needs a `make generate` and since the dependencies are now managed by dep, we need to adapt the govendor-ensure script

File: Documentation/api.md [hunk 4]
```diff
@@ -98,7 +98,7 @@ Specification of the desired behavior of the Alertmanager cluster. More info: ht
 | resources | Define resources requests and limits for single Pods. | [v1.ResourceRequirements](https://v1-6.docs.kubernetes.io/docs/api-reference/v1.6/#resourcerequirements-v1-core) | false |
 | affinity | If specified, the pod's scheduling constraints. | *v1.Affinity | false |
 | tolerations | If specified, the pod's tolerations. | []v1.Toleration | false |
-| SecurityContext | SecurityContext holds pod-level security attributes and common container settings. This defaults to non root user with uid 1000 and gid 2000. | *v1.PodSecurityContext | false |
+| securityContext | SecurityContext holds pod-level security attributes and common container settings. This defaults to non root user with uid 1000 and gid 2000. | *v1.PodSecurityContext | false |
```
Comment (ant31) [inline Documentation/api.md]: @brancz changed from `SecurityContext` to `securityContext` , is that correct?

File: Documentation/api.md [hunk 5]
```diff
@@ -63,8 +63,8 @@ AlertmanagerEndpoints defines a selection of a single Endpoints object containin
 | namespace | Namespace of Endpoints object. | string | true |
 | name | Name of Endpoints object in Namespace. | string | true |
 | port | Port the Alertmanager API is exposed on. | intstr.IntOrString | true |
-| scheme | Scheme to use when firing alerts. | string | true |
-| pathPrefix | Prefix for the HTTP path alerts are pushed to. | string | true |
+| scheme | Scheme to use when firing alerts. | string | false |
```
Comment (ant31) [inline Documentation/api.md]: Changed from required to optional

File: Documentation/api.md (hunk 4 above)
Comment (brancz) [reply Documentation/api.md]: yes! thanks :slightly_smiling_face:

File: pkg/client/monitoring/v1/types.go [hunk 6]
```diff
@@ -22,10 +22,12 @@ import (
 )
 
 // Prometheus defines a Prometheus deployment.
+// +k8s:openapi-gen=true
 type Prometheus struct {
 	metav1.TypeMeta `json:",inline"`
 	// Standard object’s metadata. More info:
 	// https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#metadata
+	// +k8s:openapi-gen=false
```
Comment (brancz) [inline pkg/client/monitoring/v1/types.go]: how come this is false?

File: scripts/govendor-ensure.sh [hunk 7]
```diff
@@ -8,12 +8,9 @@ set -u
 # print each command before executing it
 set -x
 
-go get -v github.com/kardianos/govendor
+go get -u github.com/golang/dep/cmd/dep
 
-cd vendor
-rm -r $(ls -I "vendor.json" )
-cd ..
+dep ensure
```
Comment (brancz) [inline scripts/govendor-ensure.sh]: :+1:

File: pkg/client/monitoring/v1/types.go (hunk 6 above)
Comment (ant31) [reply pkg/client/monitoring/v1/types.go]: I had an issue with this(I ll try again and document it).  I assumed that the metadata field is already validated on every resources

Comment (ant31) [reply pkg/client/monitoring/v1/types.go]: Ok so when you apply a resource with kubectl it adds a 'deletionTimestamp: ' field, but the spec requires a 'date' and doesn't allow 'nil', so the resource is rejected.

**Pull request merged**